src/kubernetes/api_watcher.rs

```diff
@@ -152,3 +164,457 @@ pub mod invocation {
+                            }{
+                                "type": "ADDED",
+                                "object": {
+                                    "kind": "Pod",
+                                    "apiVersion": "v1",
+                                    "metadata": {
+                                        "uid": "uid1"
+                                    }
+                                }
+                            }"#,
+                        );
+                }),
+                vec![
+                    Box::new(|item| {
+                        // panic! on pass is err is what we expected
+                        assert_watch_response(
+                            item,
+                            WatchEvent::Added(Pod {
+                                metadata: ObjectMeta {
+                                    uid: Some("uid0".to_owned()),
+                                    ..Default::default()
+                                },
+                                ..Default::default()
+                            }),
+                        );
+                    }),
+                    Box::new(|item| {
+                        // panic! on pass is err is what we expected
+                        assert_watch_response(
+                            item,
+                            WatchEvent::Added(Pod {
+                                metadata: ObjectMeta {
+                                    uid: Some("uid1".to_owned()),
+                                    ..Default::default()
+                                },
+                                ..Default::default()
+                            }),
+                        );
+                    }),
+                ],
+            ),
+            // Desync error at start of stream.
+            (
+                Box::new(|when, then| {
+                    when.method(GET).path("/api/v1/pods");
+                    then.status(200)
+                        .header("Content-Type", "application/json")
+                        .body(
+                            r#"{
+                                "type": "ERROR",
+                                "object": {
+                                    "apiVersion": "v1",
+                                    "code": 410,
+                                    "kind": "Status",
+                                    "message": "too old resource version: 12122 (359817167)",
+                                    "metadata": {},
+                                    "reason": "Gone",
+                                    "status": "Failure"
+                                }
+                            }"#,
+                        );
+                }),
+                vec![Box::new(|item| {
+                    // panic! on pass is err is what we expected
+                    let error = item.unwrap_err();
+                    match error {
+                        watcher::stream::Error::Desync { .. } => {}
```

Comment:
Can we assert the underlying error here? I feel like it makes a lot of sense to not just assert that this is a desync error, but to ensure the concrete variant of an error.
In addition to a more strict test, it would be kind of useful as a reference for the readers, if they ever want to unpack this error

The same applies to the rest of the assertions.